File: arrays and hashing/Rotate Array.cs
public class SolutionRotateArray
{
    public void Rotate(int[] nums, int k)
    {
        int n = nums.Length;
        k = k % n;
        int count = 0;
        for (int start = 0; count < n; start++)
        {
            int curr = start;
            int prev = nums[start];
            do
            {
                int next = (curr + k) % n;
                int temp = nums[next];
                nums[next] = prev;
                prev = temp;
                curr = next;
                count++;
            }
            while (start != curr);
        }
    }
}

File: arrays and hashing/Rotate Array.js
/**
 * @param {number[]} nums
 * @param {number} k
 * @return {void} Do not return anything, modify nums in-place instead.
 */
// var rotate = function (nums, k) {
//   let count = 0;
//   for (let start = 0; count < nums.length; start++) {
//     let current = start;
//     let prev = nums[start];
//
//     do {
//       let next = (current + k) % nums.length;
//       let temp = nums[next];
//       nums[next] = prev;
//       prev = temp;
//       current = next;
//       count++;
//     } while (current !== start);
//   }
// };

var reverse = function (left, right, arr) {
  while (left <= right) {
    let temp = arr[left];
    arr[left] = arr[right];
    arr[right] = temp;
    ++left;
    --right;
  }
};

var rotate = function (nums, k) {
  k = k % nums.length;
  reverse(0, nums.length - 1, nums);
  reverse(0, k - 1, nums);
  reverse(k, nums.length - 1, nums);
};
